src/arch/x86_64/mod.rs

```diff
@@ -111,101 +110,28 @@ pub fn get_cpu_frequency_from_os() -> std::result::Result<u32, FrequencyDetectio
 	}
 }
 
-// Constructor for a conventional segment GDT (or LDT) entry
-pub fn create_gdt_entry(flags: u64, base: u64, limit: u64) -> u64 {
-	((base & 0xff000000u64) << (56 - 24))
-		| ((flags & 0x0000f0ffu64) << 40)
-		| ((limit & 0x000f0000u64) << (48 - 16))
-		| ((base & 0x00ffffffu64) << 16)
-		| (limit & 0x0000ffffu64)
-}
-
-pub const MIN_PHYSMEM_SIZE: usize = BOOT_PDE.as_u64() as usize + 0x1000;
-
-/// Creates the pagetables and the GDT in the guest memory space.
-///
-/// The memory slice must be larger than [`MIN_PHYSMEM_SIZE`].
-/// Also, the memory `mem` needs to be zeroed for [`PAGE_SIZE`] bytes at the
-/// offsets [`BOOT_PML4`] and [`BOOT_PDPTE`], otherwise the integrity of the
-/// pagetables and thus the integrity of the guest's memory is not ensured
-pub fn initialize_pagetables(mem: &mut [u8]) {
-	assert!(mem.len() >= MIN_PHYSMEM_SIZE);
-	let mem_addr = std::ptr::addr_of_mut!(mem[0]);
-
-	let (gdt_entry, pml4, pdpte, pde);
-	// Safety:
-	// We only operate in `mem`, which is plain bytes and we have ownership of
-	// these and it is asserted to be large enough.
-	unsafe {
-		gdt_entry = mem_addr
-			.add(BOOT_GDT.as_u64() as usize)
-			.cast::<[u64; 3]>()
-			.as_mut()
-			.unwrap();
-
-		pml4 = mem_addr
-			.add(BOOT_PML4.as_u64() as usize)
-			.cast::<PageTable>()
-			.as_mut()
-			.unwrap();
-		pdpte = mem_addr
-			.add(BOOT_PDPTE.as_u64() as usize)
-			.cast::<PageTable>()
-			.as_mut()
-			.unwrap();
-		pde = mem_addr
-			.add(BOOT_PDE.as_u64() as usize)
-			.cast::<PageTable>()
-			.as_mut()
-			.unwrap();
-
-		/* For simplicity we currently use 2MB pages and only a single
-		PML4/PDPTE/PDE. */
-
-		// per default is the memory zeroed, which we allocate by the system
-		// call mmap, so the following is not necessary:
-		/*libc::memset(pml4 as *mut _ as *mut libc::c_void, 0x00, PAGE_SIZE);
-		libc::memset(pdpte as *mut _ as *mut libc::c_void, 0x00, PAGE_SIZE);
-		libc::memset(pde as *mut _ as *mut libc::c_void, 0x00, PAGE_SIZE);*/
-	}
-	// initialize GDT
-	gdt_entry[BOOT_GDT_NULL] = 0;
-	gdt_entry[BOOT_GDT_CODE] = create_gdt_entry(0xA09B, 0, 0xFFFFF);
-	gdt_entry[BOOT_GDT_DATA] = create_gdt_entry(0xC093, 0, 0xFFFFF);
-
-	pml4[0].set_addr(
-		BOOT_PDPTE,
-		PageTableFlags::PRESENT | PageTableFlags::WRITABLE,
-	);
-	pml4[511].set_addr(
-		BOOT_PML4,
-		PageTableFlags::PRESENT | PageTableFlags::WRITABLE,
-	);
-	pdpte[0].set_addr(BOOT_PDE, PageTableFlags::PRESENT | PageTableFlags::WRITABLE);
-
-	for i in 0..512 {
-		let addr = PhysAddr::new(i as u64 * Page::<Size2MiB>::SIZE);
-		pde[i].set_addr(
-			addr,
-			PageTableFlags::PRESENT | PageTableFlags::WRITABLE | PageTableFlags::HUGE_PAGE,
-		);
-	}
+pub fn init_guest_mem(mem: &mut [u8], guest_address: u64) {
+	// TODO: we should maybe return an error on failure (e.g., the memory is too small)
+	paging::initialize_pagetables(mem, guest_address);
```

Comment:
Some parts related to paging were deliberately moved into a new module because everything is getting more complex. This change is not reflected in `aarch64`, under the mantra of "change the least amount of things in `aarch64` as possible".